### DISCUSSION TITLE
Floordata injection

### DISCUSSION THREAD

**Discussion by aredfan** [LostArtefacts/TRX#755] [General]
I thought it's worth setting up a discussion group for the upcoming floordata injection feature lahm's been working on.  I've already raised two issues #750 & #751, but I thought it's best to consolidate here.  There's also the possibility of adding a flipmap trigger which would help fix the Atlantean Stronghold crash #227, that is if you're playing with the vanilla files.

With these 3 issues off the top of my head, there will probably be more to come, and so I thought it'll be a good idea to have a place to swap ideas and ask questions.  Just like the black mesh faces discussion in #736.

**Update:** List of issues related to this discussion.

- #227 
- #750 
- #751 
- #788 
- #865 
- #886 
- #939 
- #942 
- #1052 
- #1066 
- #1744 
- #1851 
- #1902 
- #1920 

__________________________

With that out of the way, I would like to draw attention to a particular switch in Obelisk of Khamoon. Apparently all Lara has to do is climb onto the block and the room get flipped.  Pulling the switch doesn't seem to do anything 🔍  

<details><summary>Image</summary>

![Screenshot (229)](https://user-images.githubusercontent.com/94776985/224470525-b56cc4a4-0bef-44de-b29f-c3770bfd0039.png)

</details>

**Comment by lahm86** [comment]
Good find in Obelisk, I'd never noticed that before. Looking at it, the issue is that the trigger in the flipped room is of the wrong type.

So when unflipped, the trigger is of type switch.
![image](https://user-images.githubusercontent.com/33758420/224477412-e882efd0-e5b6-40e2-bac4-57494b5efd8c.png)

But when flipped, it's just a regular trigger.
![image](https://user-images.githubusercontent.com/33758420/224477442-53cb5ffc-53ca-4ec8-8937-547dda40accd.png)

So that should be an easy fix :)

**Reply by aredfan** [reply]
Much appreciated, and fascinating!  I'm still figuring out how to use trview so it's actually helpful to see your explanation for this issue.

**Comment by lahm86** [comment]
In City of Khamoon, you don't need to pull the switch in room 25 to open the trapdoors, you only need to stand below them. These should be dummy triggers.
![image](https://user-images.githubusercontent.com/33758420/224477600-a09ee5df-2971-48b1-89ef-e032393e79c2.png)

Similarly, in room 34, the trapdoors will open if you stand on these tiles, so again they should be dummy triggers.
![image](https://user-images.githubusercontent.com/33758420/224477665-ad1c53e0-af90-4211-a099-909ceb94fe13.png)

**Reply by aredfan** [reply]
Agree that these two should be dummy triggers.  I did notice the first in room 25 but did not know about the second.  Nice finds 👍

**Reply by Richard-L** [reply]
On the second one: this is the room with the mummy typically attacking you in the corridor leading up to it, right?

If you're fighting the mummy there, a player wouldn't necessarily move up on those tiles I reckon. Thus you could get stuck because the doors won't open? 

Maybe I'm misremembering the level design. Not suggesting to change anything, just wrapping my head around it.

**Reply by aredfan** [reply]
Yup, this is the room with the mummy after the long corridor.  After killing the mummy you'll need to locate the switch that open the ceiling trapdoors which fills the room with sand, and thus allowing you to reach the end of the level.

However, if Lara steps on those highlighted tiles the ceiling trapdoor opens prematurely, so if you look up you'll see a black void.

**Reply by lahm86** [reply]
A potential issue with this could be with speedrunners using flipmap tricks to skip having to go to the lever. If we converted the triggers, they wouldn't be able to fall through the trapdoors. These techniques are deemed glitchless (technically exploits that are allowed).
https://youtu.be/LVkAMB0DyQk?t=208

But I know there is debate about speedrunners using Tomb1Main anyway, so it's probably not an issue at all. We can make floordata fixes optional regardless to cover all angles.

**Reply by lahm86** [reply]
An update on this one - by not changing the triggers in the flipped state we can retain the potential speedrun opportunity. The triggers will only be fixed in the unflipped state, then if you pull the switch dev-intended, the trapdoors will be open in the flipped state.

**Comment by lahm86** [comment]
Just remembered another one - in Folly, music is supposed to play when all the keys are used, but because of the masks the final bit for the music track is never set. This one might be a bit more awkward to resolve, but worth noting.

![image](https://user-images.githubusercontent.com/33758420/224480805-e2bef227-7cc4-4d72-b152-0649bb6e4b8e.png)

**Reply by Richard-L** [reply]
That's curious. What track is supposed to play?

**Reply by lahm86** [reply]
It's the same one that plays when you pick up a cog in Lost Valley (the short choral "discovery" track, not sure if it's named). Feels quite fitting. Not sure if it plays on PSX, perhaps it also has the missing trigger item.

**Reply by Richard-L** [reply]
So interesting :)

I can definitely confirm there is no music playing after the last key here on PS1.

On track naming:
Since Nathan McCree's music is so wonderful, I've spent a bit more time with the soundtracks of TR1-3. The names of TR3's tracks are contained in the `.WAD` file, but TR1 and 2 used audio tracks on CD of course. This will then allow you to have the CDs metadata fetched from online music databases. In my case I've done this with Apple's iTunes, which accesses the Gracenote database. I've then mounted the different region releases of the PS1 and Saturn discs, to see what could be obtained. Not all of them matched a record, but of those that did there were sometimes different track names fetched across releases:

![image](https://user-images.githubusercontent.com/4934209/224546764-c1c799ef-650a-4156-a475-0695b41ac0e3.png)

Where there've been two track names, I've included them as "name 1 / name 2". I remember I had posted about this here as well https://github.com/LostArtefacts/Tomb1Main/discussions/247#discussioncomment-1678769

The ones in Lost Valley are:
- Calm III / The Sacred Place (approaching the bridge jump)
- Calm II / A Discovery (after the bridge, picking up the cog)
- Calm IV / The Evil Place (pulling the lever with all cogs inserted)

**Reply by lahm86** [reply]
That's really interesting - nice investigation :)
This would be the proposed result for Folly. It also fixes the ambient track being stopped when the first key is used. This is direct level editing at the moment, I haven't started on the actual FD injection.
https://www.youtube.com/watch?v=WFbHQLzMPIQ

**Reply by Richard-L** [reply]
That's really excellent if you ask me. Very believable this was the intended effect.

I'm so glad all these little discoveries can be touched upon little by little. Never dreamed the game could be elevated to these heights.

**Comment by Rapora9** [comment]
Talking of City of Khamun, I would like to see this mummy return to PC version as well. The mummy is there on PlayStation and Sega Saturn, but apparently not on Japanese PS version.

![Näyttökuva (247)](https://user-images.githubusercontent.com/97730054/224542297-9cd3724d-7d9c-4dfc-a71f-53d1b5e7b413.png)

I'm not sure of its exact location and trigger so they should be checked by someone. Also, I do understand this mummy position might be annoying for people, so some would probably want to not have this mummy around, while also being fine with other floordata changes.

**Reply by lahm86** [reply]
We restored this mummy in the randomizer, but guessed at its position and trigger position tbh. So a PlayStation expert would be great for input here :) I guess the mummy isn't visible when you're below the bridge?

![image](https://user-images.githubusercontent.com/33758420/224545112-98e2f171-1b86-4c49-9dea-e122bb688beb.png)

I agree too this should be a separate option as it changes the enemy count. I wonder if there are any other instances of this in other levels?

**Reply by Rapora9** [reply]
The trigger seems to be on the starting position of that pushable (because as you can see, the mummy is active already when Lara is standing on that spot on my screenshot).

I didn't see the mummy from below in the video I was watching:

![Näyttökuva (248)](https://user-images.githubusercontent.com/97730054/224545781-4babe66e-7c82-4e79-b33a-999f455ec9c2.png)

**Comment by aredfan** [comment]
In Temple of the Cat some players have found a way to survive the skull room and gain the extra kills.  However the flipmap trigger in the skull room changes the next area, meaning the wall bars disappear and the giant cat's tongue gets extended.

<details><summary>Image</summary>

![Screenshot (237)](https://user-images.githubusercontent.com/94776985/224805918-439bbdf1-6cf1-4b9b-84e7-c23ae2af32c3.png)

</details>

I think it would be a good idea to add flipmap triggers to the rooms on either side of the skull room so that the next room gets unflipped again.  This would restore the proper way of completing the next area i.e. restore the bars and retract the giant cat's tongue.

<details><summary>Image</summary>

![Screenshot (238)](https://user-images.githubusercontent.com/94776985/224806539-b3ccf954-42b6-493d-ade3-15627615f148.png)

</details>

**Reply by lahm86** [reply]
We feel that it's best not to implement this one as the only gain would be a speedrun-barrier rather than impacting dev-intended play. It differs from the City of Khamoon issue above as the fix there provides a visual fix in the unflipped state.

**Comment by aredfan** [comment]
I've come across a switch in The Hive level which doesn't seem to do anything.  From reading Stella's guide this switch is meant to reopen the timed doors and allow Lara to backtrack if needed.

<details><summary>Image</summary>

![Screenshot](https://user-images.githubusercontent.com/94776985/226098168-dd3aca9e-494d-4da7-b74d-322a28fab51e.png)

</details>

**Reply by aredfan** [reply]
Upon further investigation this is a bit of a weird switch that depends on the the order you do it in.

- If you've stepped on one of the timed-door triggers without stepping on the other, the switch will open the other door.
- If you've stepped on both timed-door triggers, the switch won't open either doors.

**Reply by lahm86** [reply]
This one is tricky as it would involve a lot of internal changes to timed and switch triggers and can't be fixed immediately with the FD manipulation we support. It is possible to get back out of the room so I think we should leave it as-is.

**Reply by aredfan** [reply]
No problem at all!  This issue was more of a curiosity for me but I'm more than happy to go with your decision in this case 👍

**Comment by walkawayy** [comment]
Non one-shot music triggers that should be changed to one-shot:

**Reply by walkawayy** [reply]
Caves:
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/ebbe8ac2-807b-4847-86fa-24936af02657)

![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/3d1e291b-90e3-4e69-9667-c5e6148d953b)

**Reply by walkawayy** [reply]
Vilcabamba:
None

**Reply by walkawayy** [reply]
Lost Valley:
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/7b445324-1e22-4870-9939-45d388202f8b)

**Reply by walkawayy** [reply]
Qualapec:
I think this one for sure. Trigger above it is an antipad. So I think it needs to stay (similar to switches). Lmk if that's wrong.
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/de3144a5-fc1b-4ec9-a994-0f5118d50951)

**Reply by walkawayy** [reply]
St. Francis Folly:
This one replays music. But idk if it can be changed because of the flip off.
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/cd4795d0-52c3-4683-9a03-19df62ec33a6)

**Reply by walkawayy** [reply]
Colosseum:
None

**Reply by walkawayy** [reply]
Palace Midas:
None

**Reply by walkawayy** [reply]
Cistern:
None

**Reply by walkawayy** [reply]
Tihocan:
None

**Reply by walkawayy** [reply]
City of Khamoon:
None

**Reply by walkawayy** [reply]
Obelisk of Khamoon:
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/c4e6882c-8c7c-4a02-a753-9ea66484eb43)

![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/b8c86c78-f22d-474c-94f3-f463d23334fe)

![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/5142de4c-ab61-4c15-b6ed-06859b0dd24f)

**Reply by walkawayy** [reply]
Sanctuary of the Scion:
all 3
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/cc2d3094-a37f-4961-9f91-c83785b1e75b)

all 6
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/d7a59c23-c144-497f-88ef-7198e7268f5f)

**Reply by walkawayy** [reply]
Natla's Mines:
2
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/a4b5f511-40d0-47db-9721-6a6fe58c1133)

**Reply by walkawayy** [reply]
Atlantis:
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/e636efb6-3faa-4141-b4cd-88c4bd7a74fb)

**Reply by walkawayy** [reply]
Great Pyramid:
![image](https://github.com/LostArtefacts/Tomb1Main/assets/81546780/352a7af1-3d6b-4bd4-b3ef-c1edd5515a81)

**Reply by aredfan** [reply]
I've verified all your finds @walkawayy, in case you needed a fresh pair of eyes.

Correct me if I'm wrong but I think you missed this one, in Palace Midas:

Edit: Whoops I meant Colosseum

<details><summary>Image</summary>

![Screenshot (451)](https://github.com/LostArtefacts/Tomb1Main/assets/94776985/070ac9cc-145c-4759-8a04-9360280f4b47)

</details>

**Reply by aredfan** [reply]
TRUB with music triggers that have `Only once` set to `false`.  I figured it's best to do a list of # otherwise it'll be a very screenshot-heavy reply.

**Return to Egypt**

- 2/5 (Flip)
- 26 (Flip)
- 35 (Flip)
- 38/39/40/41
- 134 (Flip)
- 215 

**Temple of the Cat**

- 16/17
- 266
- 269/270/271

**Atlantean Stronghold**

- 49/50/51/52
- 120 (Flip)
- 122/123
- 145
- 149 (Flip)
- 237

**The Hive**

- 124
- 258/266/273
- 298/299/300
- 301/302
- 303
- 304/305
- 311
- 325 to 345

**Reply by walkawayy** [reply]
Thanks! I wasn't sure how to handle UB because the triggers are fan made and there's so many that seem messed up. I'm also still not totally sure how to handle multi trigger tiles. In the list I did, I ignored non one shot triggers like pickups with music since Lara can obviously only get the item once. Pads I was more unsure of and switches too.

**Reply by Richard-L** [reply]
I don't fully understand what these injections are about, but just NB that the fanmade music triggers differ per release. There's the one by Ghostblade + Jonathan, and then the one by Sardoc (what TombATI uses).

I don't know which one you're basing all of this on, but probably the latter.

**Reply by aredfan** [reply]
Yup, we were referring to Sardoc's TRUB mod with added music triggers.

**Reply by lahm86** [reply]
Are the triggers positioned differently between these versions, or do they just play different tracks?

**Reply by Richard-L** [reply]
I don't know but if you wanted to check:

~~https://www.tombraiderforums.com/showthread.php?t=194134~~

https://www.tombraiderforums.com/showthread.php?t=213223

**Reply by lahm86** [reply]
Ok, so the Ghostblade/Jonathan thread is for the PSX version, is that right? Is there only one PC version of TRUB?

**Reply by walkawayy** [reply]
AFAIK. There's also this version with the Atlantean Stronghold bug fixed but no music. Also from Sardoc.

https://www.tombraiderforums.com/showpost.php?p=8010339&postcount=2

**Reply by lahm86** [reply]
I think we'll just have to safeguard as follows with these particular injections:
* Look for a trigger in room N at position X,Y
  * If it has a music action, convert to OneShot

Will also test each case where there are other actions to ensure behaviour is as expected.

Anyone using non-music files won't have triggers that serve other purposes change. For any other potential levels out there that have triggers in different positions, I can't really see it viable to provide injections - my feeling is it should just be the standard that the T1M installer downloads. In the future when there is a standalone injection tool, that could be an option for fans. What do you think?

**Reply by Richard-L** [reply]
Sorry Lahm I got something confused 🙏

There _are_ two UB versions, but the other next to Sardoc's is retail, and it has no music triggers, thus is irrelevant for this discussion.

So my first link above is wrong.

**Reply by lahm86** [reply]
No worries at all.
I'm going to make a start on the FD injections soon, will begin at the top of the thread so if there is anything else to note just let us know :)

**Comment by aredfan** [comment]
Closing this discussion as it's been inactive for more than a year.  We can instead create issues for new suggestions/requests that may arise in the future.
